### PR TITLE
Emit FSTRING_MIDDLE for literal after last expression and before quotes

### PR DESCRIPTION
**Pull request by lysnikolaou**
cc @pablogsal @isidentical

### CODE REVIEW

File: Parser/tokenizer.c [hunk 1]
```diff
@@ -2437,18 +2437,45 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
...
         return tok_get_normal_mode(tok, current_tok, token);
     }
 
+    // Emit FSTRING_END in case we've reached the end of the string
+    if ((current_tok->f_string_quote_size == 1 && start_char == current_tok->f_string_quote)
+            || (current_tok->f_string_quote_size == 3 && start_char == current_tok->f_string_quote
+                && peek1 == current_tok->f_string_quote && peek2 == current_tok->f_string_quote)) {
+
+        // Advance the tokenizer state again to create a token out of the end quotes
+        tok_nextc(tok);
```
Comment (pablogsal) [inline Parser/tokenizer.c:2460]: Nit (feel free to ignore): Here we can probably just loop over `f_string_quote_size` and call `tok_nextc` in the body of the loop

File: Parser/tokenizer.c [hunk 2]
```diff
@@ -2437,18 +2437,45 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
...
             current_tok->bracket_mark[++current_tok->bracket_mark_index] = current_tok->bracket_stack;
         }
         tok->tok_mode_stack[tok->tok_mode_stack_index].kind = TOK_REGULAR_MODE;
         return tok_get_normal_mode(tok, current_tok, token);
     }
 
+    // Emit FSTRING_END in case we've reached the end of the string
+    if ((current_tok->f_string_quote_size == 1 && start_char == current_tok->f_string_quote)
+            || (current_tok->f_string_quote_size == 3 && start_char == current_tok->f_string_quote
+                && peek1 == current_tok->f_string_quote && peek2 == current_tok->f_string_quote)) {
```
Comment (pablogsal) [inline Parser/tokenizer.c:2457]: You can compare against `start_char` after `start_char == current_tok->f_string_quote` which is shorter. Also I think this whole thing can be simplified to:

```
    if (start_char == current_tok->f_string_quote && (current_tok->f_string_quote_size != 3  || (start_char ==  peek1 
 && start_char == peel2))) {
```

File: Parser/tokenizer.c [hunk 3]
```diff
@@ -2550,17 +2577,16 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
...
     }
 
-    if (current_tok->last_expr_buffer != NULL) {
-        PyMem_Free(current_tok->last_expr_buffer);
-        current_tok->last_expr_buffer = NULL;
-        current_tok->last_expr_size = 0;
-        current_tok->last_expr_end = -1;
+    // Backup the f-string quotes to emit a final FSTRING_MIDDLE and
+    // add the quotes to the FSTRING_END in the next tokenizer iteration.
+    tok_backup(tok, current_tok->f_string_quote);
```
Comment (pablogsal) [inline Parser/tokenizer.c:2582]: Same as before, we can probably just loop over `f_string_quote_size`

**Review by isidentical — COMMENTED** [review]
aside from the Pablo's comments, I think this look very good (and clever) 👍🏻

**Comment by pablogsal** [comment]
@lysnikolaou Will land as soon as you tell me you are happy with the current status.

**Comment by mgmacias95** [comment]
Running #46  + #51 + this PR produces this result in the tokenizer test (`./python.exe -m test test_tokenize -v `):

```
======================================================================
FAIL: test_string (test.test_tokenize.CTokenizeTest.test_string) (source='Rf"abc\\\ndef"')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/mgmacias/Documents/GitHub/cpython/Lib/test/test_tokenize.py", line 1666, in check_tokenize
    self.assertEqual(result, expected.rstrip().splitlines())
AssertionError: Lists differ: ['   [55 chars]RING_MIDDLE 'abc\\\\\\ndef'  (1, 3) (2, 3)", '[43 chars] 4)'] != ['   [55 chars]RING_END 'abc\\\\\\ndef'  (2, -1) (2, 3)"]

First differing element 1:
"    FSTRING_MIDDLE 'abc\\\\\\ndef'  (1, 3) (2, 3)"
"    FSTRING_END 'abc\\\\\\ndef'  (2, -1) (2, 3)"

First list contains 1 additional elements.
First extra element 2:
'    FSTRING_END \'"\'           (2, 11) (2, 4)'

  ['    FSTRING_START \'Rf"\'         (1, 0) (1, 3)',
-  "    FSTRING_MIDDLE 'abc\\\\\\ndef'  (1, 3) (2, 3)",
?               ^^ ---                  -------       ^

+  "    FSTRING_END 'abc\\\\\\ndef'  (2, -1) (2, 3)"]
?               ^^                       ++++++++   ^

-  '    FSTRING_END \'"\'           (2, 11) (2, 4)']

----------------------------------------------------------------------
```

That (2,11) tuple looks like a bug when getting the line number.

**Comment by pablogsal** [comment]
I merged both PRs so you can rebase this on top of my branch to see the failures

**Comment by lysnikolaou** [comment]
@mgmacias95 Thanks for the report. It was really helpful.

@lysnikolaou @isidentical I think this is ready for another review.

File: Include/internal/pycore_token.h [hunk 4]
```diff
@@ -91,8 +91,7 @@ extern "C" {
                                  (x) == INDENT    || \
                                  (x) == DEDENT)
 #define ISSTRINGLIT(x)          ((x) == STRING           || \
-                                 (x) == FSTRING_MIDDLE   || \
-                                 (x) == FSTRING_END)
+                                 (x) == FSTRING_MIDDLE)
```
Comment (isidentical) [inline Include/internal/pycore_token.h:94]: @lysnikolaou I think this file is auto-generated so we might need to update the generate_tokens.py

Comment (lysnikolaou) [reply Include/internal/pycore_token.h:94]: Oh, that's correct! Completely missed it. I've updated that as well.

**Review by isidentical — APPROVED** [review]
love the simplification on the logic 💯 let's land this!

**Review by mgmacias95 — APPROVED** [review]
LGTM

**Review by pablogsal — APPROVED** [review]
Φανταστική δουλειά!

**Pull request merged**